depthai_ros_driver/test/conversion.cpp

```diff
@@ -0,0 +1,109 @@
+#include <depthai_ros_driver/conversion.hpp>
+
+#include <gtest/gtest.h>
+
+namespace rr {
+bool cvMatAreEqual(const cv::Mat& a, const cv::Mat& b) {
+    if (a.size() != b.size()) {
+        return false;
+    }
+
+    cv::Mat temp;
+    cv::bitwise_xor(a, b, temp);
+    return cv::countNonZero(temp.reshape(1)) == 0;
+}
+
+void iota_fill(uchar* data, int n) {
+    for (int i = 0; i < n; ++i) {
+        data[i] = i;
+    }
+}
+
+TEST(conversion, base_comparison) {
+    // test that cvMatAreEqual is working as expected
+    uchar* a = new uchar[3]{1, 2, 3};
+    uchar* b = new uchar[3]{1, 2, 3};
+    uchar* c = new uchar[3]{4, 5, 6};
+
+    cv::Mat A(3, 1, CV_8UC1, a);
+    cv::Mat B(3, 1, CV_8UC1, b);
+    cv::Mat C(3, 1, CV_8UC1, c);
+
+    EXPECT_TRUE(cvMatAreEqual(A, B));
+    EXPECT_FALSE(cvMatAreEqual(B, C));
+    EXPECT_FALSE(cvMatAreEqual(A, C));
+
+    delete[] a;
+    delete[] b;
+    delete[] c;
+}
+
+TEST(conversion, base_fill) {
+    // test that iota_fill is working as expected
+    cv::Mat A(1, 2, CV_8UC3);
+
+    iota_fill(A.data, A.cols * A.rows * A.channels());
+
+    auto pt1 = A.at<cv::Vec3b>(0);
+    auto pt2 = A.at<cv::Vec3b>(1);
+
+    EXPECT_EQ(pt1[0], 0);
+    EXPECT_EQ(pt1[1], 1);
+    EXPECT_EQ(pt1[2], 2);
+
+    EXPECT_EQ(pt2[0], 3);
+    EXPECT_EQ(pt2[1], 4);
+    EXPECT_EQ(pt2[2], 5);
+}
+
+TEST(conversion, planar2interleaved2planar_3x3) {
+    cv::Mat planar_source(3, 3, CV_8UC3);
+    uchar* src = planar_source.data;
+    iota_fill(src, planar_source.rows * planar_source.cols * planar_source.channels());
```

Comment:
it is okay now, but this may not work in the future when `planar2interleaved` is optimized using `cv::merge` etc.
OpenCV can have an implementation that the data address of each row has the address multiples of e.g. 16. Just for future information.

Reply:
![mat_memory](https://user-images.githubusercontent.com/49298092/127452695-df7f7fbf-8e2e-45bf-8f38-a5aadb1ef183.png)

There seems no padding between rows. cling is so useful.